Implement cached sequence number

lvn-hasky-dragon opened the pull request.

lvn-hasky-dragon committed:
Implement cached sequence number

lvn-hasky-dragon committed:
Update transaction by one

lvn-hasky-dragon committed:
cargo fmt

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,74 @@ impl<Res> PerformQueryWrapper<Res> {
+    ) -> Result<BaseAccount, Error> {
+        let mut guard = self.pool.get().await?;
+        let cosmos = guard.get_inner_mut();
+        let mut sequences = cosmos.simulate_sequences.write().await;
```
This approach means we're grabbing a write lock and holding onto it while querying the chain, which will likely block other tasks fairly often. It would be best to avoid that.

lvn-hasky-dragon replied:
Updated.

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,74 @@ impl<Res> PerformQueryWrapper<Res> {
 }
 
 impl Cosmos {
+    async fn get_and_update_simulation_sequence(
```
I'm not sure we need the simulation sequence, but I could be mistaken. My theory is that the same node (or cluster of nodes) that we're requesting the current sequence number from will also be doing the simulation. The place where the problem really should be occurring is when the node we're simulating on disagrees with the current state of the chain (and therefore the current validator).

Maybe an important next step in this is modifying the error reporting for health reports to tell us whether the sequence mismatch occurred during the simulate or broadcast phase.

lvn-hasky-dragon replied:
I will create a new PR for updating the health report tomorrow to track where the failure is happening. Right now I think we have two options:
- Leave this as it is
- Remove this function

Don't have strong opinion either wise. What would you prefer ?

lvn-rusty-dragon replied:
We can leave it as-is for testing. We can deploy some testnet bots with all these changes and see what happens, it should be enlightening.

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,74 @@ impl<Res> PerformQueryWrapper<Res> {
+        let mut sequences = cosmos.broadcast_sequences.write().await;
+        sequences
+            .entry(address)
+            .and_modify(|item| item.sequence += 1);
```
This looks a bit dangerous to me. We just broadcast a transaction, so we know the sequence number we used. Instead of incrementing the cache, which may be out of sync with the transaction we just made, I'd recommend taking it from the transaction itself.

lvn-hasky-dragon replied:
Sure, I'm using Tx now.

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,74 @@ impl<Res> PerformQueryWrapper<Res> {
+        &self,
+        address: Address,
+    ) -> Result<BaseAccount, Error> {
+        let mut guard = self.pool.get().await?;
```
Same locking comment as above.

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,74 @@ impl<Res> PerformQueryWrapper<Res> {
+            });
+        let mut base_account = self.get_base_account(address).await?;
+        let diff = Instant::now().duration_since(sequence_info.timestamp);
+        if diff.as_secs() <= 30 {
```
If the diff is greater than 30 seconds, shouldn't we store the new value in the cache?

lvn-hasky-dragon replied:
Oops, yeah I totally missed it. Have included a fix.

lvn-hasky-dragon committed:
Update based on lvn-rusty-dragon's comments

lvn-hasky-dragon committed:
Modify health report to show during which action it failed

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,131 @@ impl<Res> PerformQueryWrapper<Res> {
+    ) -> Result<BaseAccount, Error> {
+        let mut guard = self.pool.get().await?;
+        let cosmos = guard.get_inner_mut();
+        let sequences = cosmos.simulate_sequences.read().await;
```
Unless I'm mistaken: this is going to take a read guard for the rest of this function call, which will create a deadlock when we try to take the write guard below. This read guard should be dropped immediately.

lvn-hasky-dragon replied:
You're right. Have pushed a fix for it.

lvn-rusty-dragon commented on packages/cosmos/src/client.rs:
```diff
@@ -147,6 +148,131 @@ impl<Res> PerformQueryWrapper<Res> {
+            let sequence = &auth_info
+                .signer_infos
+                .iter()
+                .map(|item| item.sequence)
```
This only works right now since we have a limitation of only allowing a single signer per transaction. Probably worth at least a comment about this.

lvn-hasky-dragon replied:
Added a comment.

lvn-hasky-dragon committed:
Drop read guards in scope

lvn-rusty-dragon reviewed (approved):
Like the previous PR: should we keep this as a PR and do some testing, or do you want to merge into develop? I'm comfortable either way, your call.

lvn-hasky-dragon committed:
Implement account sequence mismsatch for broadcast

lvn-hasky-dragon committed:
Rename to indicate reality

lvn-hasky-dragon committed:
Fix clippy issue

lvn-hasky-dragon committed:
Logging on successful account sequence mismatch retry

Right now, I'm unable to understand if the new patches is actually
improving thing.

lvn-hasky-dragon committed:
Merge pull request #54 from Levana-Protocol/broadcast-seq-no

Retry with expected sequence number on simulation when it fails because of account sequence mismatch issue

lvn-hasky-dragon committed:
Print on failurs too

lvn-hasky-dragon committed:
Fix typo

lvn-hasky-dragon committed:
Merge pull request #55 from Levana-Protocol/info-bsn

Logging on successful account sequence mismatch retry
